File: bench/chorin_gs_bench.c
/*
 * FlowLabLite — Chorin + Gauss-Seidel benchmark  (C)
 *
 * Matches FlowLabLite v0.0.1 Chorin solver (tag v0.0.1,
 * cmd/main/main.mbt cavity_flow_array).
 *
 * Compile & run:
 *   gcc -O2 -o chorin_gs_bench chorin_gs_bench.c -lm
 *   ./chorin_gs_bench              # all grid sizes
 *   ./chorin_gs_bench 41 500 5    # n nt repeats
 *
 * Physical parameters: domain 2×2, rho=1.0, nu=0.1 (Re=20), dt=0.001
 * BCs: top lid u=1, all walls no-slip
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

#define RHO  1.0
#define NU   0.1
#define DT   0.001
#define NIT  50
#define L    2.0

/* Wall-clock timer in milliseconds */
static double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1.0e6;
}

/* Allocate n×n double array (flat row-major) */
static double* alloc2d(int n) {
    double* a = (double*)calloc(n * n, sizeof(double));
    if (!a) { fprintf(stderr, "OOM\n"); exit(1); }
    return a;
}

#define IDX(n, i, j)  ((i)*(n) + (j))
#define U(i,j)  u[IDX(n,i,j)]
#define V(i,j)  v[IDX(n,i,j)]
#define P(i,j)  p[IDX(n,i,j)]
#define UN(i,j) un[IDX(n,i,j)]
#define VN(i,j) vn[IDX(n,i,j)]
#define PN(i,j) pn[IDX(n,i,j)]
#define B(i,j)  b[IDX(n,i,j)]

static void solve(int n, int nt) {
    double dx  = L / (n - 1);
    double dy  = L / (n - 1);
    double dx2 = dx * dx;
    double dy2 = dy * dy;
    double invD = 1.0 / (2.0 * (dx2 + dy2));
    double coef = dx2 * dy2 * invD;

    double *u  = alloc2d(n), *v  = alloc2d(n), *p  = alloc2d(n);
    double *un = alloc2d(n), *vn = alloc2d(n), *pn = alloc2d(n);
    double *b  = alloc2d(n);

    for (int step = 0; step < nt; step++) {

        /* Save old velocities */
        memcpy(un, u, n*n*sizeof(double));
        memcpy(vn, v, n*n*sizeof(double));

        /* Build RHS b */
        for (int i = 1; i < n-1; i++)
            for (int j = 1; j < n-1; j++)
                B(i,j) = RHO/DT * (
                    (UN(i,j+1) - UN(i,j-1)) / (2*dx) +
                    (VN(i+1,j) - VN(i-1,j)) / (2*dy)
                );

        /* Gauss-Seidel pressure Poisson */
        for (int q = 0; q < NIT; q++) {
            memcpy(pn, p, n*n*sizeof(double));
            for (int i = 1; i < n-1; i++)
                for (int j = 1; j < n-1; j++) {
                    double lap = (PN(i,j+1)+PN(i,j-1))*dy2
                               + (PN(i+1,j)+PN(i-1,j))*dx2;
                    P(i,j) = lap * invD - coef * B(i,j);
                }
            /* BCs */
            for (int i = 0; i < n; i++) P(i,n-1) = P(i,n-2);
            for (int j = 0; j < n; j++) P(0,j)   = P(1,j);
            for (int i = 0; i < n; i++) P(i,0)   = P(i,1);
            for (int j = 0; j < n; j++) P(n-1,j) = 0.0;
        }

        /* Velocity update */
        for (int i = 1; i < n-1; i++)
            for (int j = 1; j < n-1; j++) {
                double cux = UN(i,j)*DT/dx*(UN(i,j)-UN(i,j-1));
                double cuy = VN(i,j)*DT/dy*(UN(i,j)-UN(i-1,j));
                double pgx = DT/(2*RHO*dx)*(P(i,j+1)-P(i,j-1));
                double vux = NU*DT/dx2*(UN(i,j+1)-2*UN(i,j)+UN(i,j-1));
                double vuy = NU*DT/dy2*(UN(i+1,j)-2*UN(i,j)+UN(i-1,j));
                U(i,j) = UN(i,j) - cux - cuy - pgx + vux + vuy;

                double cvx = UN(i,j)*DT/dx*(VN(i,j)-VN(i,j-1));
                double cvy = VN(i,j)*DT/dy*(VN(i,j)-VN(i-1,j));
                double pgy = DT/(2*RHO*dy)*(P(i+1,j)-P(i-1,j));
                double vvx = NU*DT/dx2*(VN(i,j+1)-2*VN(i,j)+VN(i,j-1));
                double vvy = NU*DT/dy2*(VN(i+1,j)-2*VN(i,j)+VN(i-1,j));
                V(i,j) = VN(i,j) - cvx - cvy - pgy + vvx + vvy;
            }

        /* BCs */
        for (int j = 0; j < n; j++) U(0,j) = 0.0;
        for (int i = 0; i < n; i++) { U(i,0) = 0.0; U(i,n-1) = 0.0; }
        for (int j = 0; j < n; j++) U(n-1,j) = 1.0;
        for (int j = 0; j < n; j++) { V(0,j) = 0.0; V(n-1,j) = 0.0; }
        for (int i = 0; i < n; i++) { V(i,0) = 0.0; V(i,n-1) = 0.0; }
    }

    /* Prevent optimiser from eliminating computation */
    if (isnan(U(n/2, n/2))) fprintf(stderr, "NaN\n");

    free(u); free(v); free(p);
    free(un); free(vn); free(pn); free(b);
}

int main(int argc, char *argv[]) {
    typedef struct { int n, nt, reps; const char *label; } Case;
    Case cases[] = {
        { 41,  500, 5, "small  (41×41)"  },
        { 81,  500, 3, "medium (81×81)"  },
        { 161, 500, 3, "large (161×161)" },
    };
    int ncases = 3;

    if (argc >= 3) {
        cases[0].n    = atoi(argv[1]);
        cases[0].nt   = atoi(argv[2]);
        cases[0].reps = argc >= 4 ? atoi(argv[3]) : 3;
        cases[0].label = "custom";
        ncases = 1;
    }

    printf("%-70s\n", "FlowLabLite — Chorin+GS C benchmark (gcc -O2)");
    printf("%-70s\n\n", "=");

    printf("=== TSV_RESULTS ===\n");
    printf("grid\tnt\tc_ms\n");

    for (int ci = 0; ci < ncases; ci++) {
        Case *c = &cases[ci];
        double best = 1e18, total = 0;
        for (int r = 0; r < c->reps; r++) {
            double t0 = now_ms();
            solve(c->n, c->nt);
            double t1 = now_ms();
            double elapsed = t1 - t0;
            if (elapsed < best) best = elapsed;
            total += elapsed;
        }
        double avg = total / c->reps;
        printf("  %-22s  n=%4d  nt=%5d  best=%9.1f ms  avg=%9.1f ms\n",
               c->label, c->n, c->nt, best, avg);
        printf("%d×%d\t%d\t%.1f\n", c->n, c->n, c->nt, best);
    }
    return 0;
}

File: bench/ChorinGSBench.java
/**
 * FlowLabLite — Chorin + Gauss-Seidel benchmark  (Java)
 *
 * Matches FlowLabLite v0.0.1 Chorin solver (tag v0.0.1,
 * cmd/main/main.mbt cavity_flow_array).
 *
 * Physical parameters: domain 2×2, rho=1.0, nu=0.1 (Re=20), dt=0.001
 * BCs: top lid u=1, all walls no-slip
 *
 * Build & run:
 *   javac ChorinGSBench.java
 *   java -server ChorinGSBench            # default: all grid sizes
 *   java -server ChorinGSBench 41 500 5   # n nt repeats
 */
public class ChorinGSBench {

    static final double RHO = 1.0;
    static final double NU  = 0.1;
    static final double DT  = 0.001;
    static final int    NIT = 50;
    static final double L   = 2.0;

    // ── Core solver ─────────────────────────────────────────────────────────

    static void solve(int n, int nt) {
        double dx  = L / (n - 1);
        double dy  = L / (n - 1);
        double dx2 = dx * dx;
        double dy2 = dy * dy;
        double invD = 1.0 / (2.0 * (dx2 + dy2));
        double coef = dx2 * dy2 * invD;

        double[][] u  = new double[n][n];
        double[][] v  = new double[n][n];
        double[][] p  = new double[n][n];
        double[][] un = new double[n][n];
        double[][] vn = new double[n][n];
        double[][] b  = new double[n][n];
        double[][] pn = new double[n][n];

        for (int step = 0; step < nt; step++) {

            // Save old velocities
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++) { un[i][j] = u[i][j]; vn[i][j] = v[i][j]; }

            // Build RHS b
            for (int i = 1; i < n - 1; i++)
                for (int j = 1; j < n - 1; j++)
                    b[i][j] = RHO / DT * (
                        (un[i][j+1] - un[i][j-1]) / (2*dx) +
                        (vn[i+1][j] - vn[i-1][j]) / (2*dy)
                    );

            // Gauss-Seidel pressure Poisson
            for (int q = 0; q < NIT; q++) {
                for (int i = 0; i < n; i++)
                    for (int j = 0; j < n; j++) pn[i][j] = p[i][j];

                for (int i = 1; i < n - 1; i++)
                    for (int j = 1; j < n - 1; j++) {
                        double lap = (pn[i][j+1] + pn[i][j-1]) * dy2
                                   + (pn[i+1][j] + pn[i-1][j]) * dx2;
                        p[i][j] = lap * invD - coef * b[i][j];
                    }
                // BCs
                for (int i = 0; i < n; i++) p[i][n-1] = p[i][n-2];
                for (int j = 0; j < n; j++) p[0][j]   = p[1][j];
                for (int i = 0; i < n; i++) p[i][0]   = p[i][1];
                for (int j = 0; j < n; j++) p[n-1][j] = 0.0;
            }

            // Velocity update
            for (int i = 1; i < n - 1; i++)
                for (int j = 1; j < n - 1; j++) {
                    double cux = un[i][j] * DT/dx * (un[i][j] - un[i][j-1]);
                    double cuy = vn[i][j] * DT/dy * (un[i][j] - un[i-1][j]);
                    double pgx = DT / (2*RHO*dx) * (p[i][j+1] - p[i][j-1]);
                    double vux = NU * DT/dx2 * (un[i][j+1] - 2*un[i][j] + un[i][j-1]);
                    double vuy = NU * DT/dy2 * (un[i+1][j] - 2*un[i][j] + un[i-1][j]);
                    u[i][j] = un[i][j] - cux - cuy - pgx + vux + vuy;

                    double cvx = un[i][j] * DT/dx * (vn[i][j] - vn[i][j-1]);
                    double cvy = vn[i][j] * DT/dy * (vn[i][j] - vn[i-1][j]);
                    double pgy = DT / (2*RHO*dy) * (p[i+1][j] - p[i-1][j]);
                    double vvx = NU * DT/dx2 * (vn[i][j+1] - 2*vn[i][j] + vn[i][j-1]);
                    double vvy = NU * DT/dy2 * (vn[i+1][j] - 2*vn[i][j] + vn[i-1][j]);
                    v[i][j] = vn[i][j] - cvx - cvy - pgy + vvx + vvy;
                }

            // BCs
            for (int j = 0; j < n; j++) u[0][j] = 0.0;
            for (int i = 0; i < n; i++) { u[i][0] = 0.0; u[i][n-1] = 0.0; }
            for (int j = 0; j < n; j++) u[n-1][j] = 1.0;
            for (int j = 0; j < n; j++) { v[0][j] = 0.0; v[n-1][j] = 0.0; }
            for (int i = 0; i < n; i++) { v[i][0] = 0.0; v[i][n-1] = 0.0; }
        }
        // Prevent JIT from eliminating the computation
        if (Double.isNaN(u[n/2][n/2])) System.err.println("NaN detected");
    }

    // ── Benchmark runner ─────────────────────────────────────────────────────

    record GridCase(int n, int nt, int repeats, String label) {}

    static final GridCase[] CASES = {
        new GridCase(41,  500, 5, "small  (41×41)"),
        new GridCase(81,  500, 3, "medium (81×81)"),
        new GridCase(161, 500, 3, "large (161×161)"),
    };

    public static void main(String[] args) {
        System.out.println("=".repeat(70));
        System.out.println("FlowLabLite — Chorin+GS Java benchmark (JDK " +
            System.getProperty("java.version") + ")");
        System.out.println("=".repeat(70));

        GridCase[] cases;
        if (args.length >= 2) {
            int n  = Integer.parseInt(args[0]);
            int nt = Integer.parseInt(args[1]);
            int rp = args.length >= 3 ? Integer.parseInt(args[2]) : 3;
            cases  = new GridCase[]{ new GridCase(n, nt, rp, n + "×" + n) };
        } else {
            cases = CASES;
        }

        // JVM warmup (run smallest case twice before timing)
        System.out.println("JVM warmup...");
        solve(cases[0].n(), cases[0].nt());
        solve(cases[0].n(), cases[0].nt());
        System.out.println();

        System.out.println("=== TSV_RESULTS ===");
        System.out.println("grid\tnt\tjava_ms");

        for (GridCase c : cases) {
            double[] times = new double[c.repeats()];
            for (int r = 0; r < c.repeats(); r++) {
                long t0 = System.nanoTime();
                solve(c.n(), c.nt());
                long t1 = System.nanoTime();
                times[r] = (t1 - t0) / 1_000_000.0;
            }
            double best = Double.MAX_VALUE, sum = 0;
            for (double t : times) { if (t < best) best = t; sum += t; }
            double avg = sum / times.length;
            System.out.printf("  %-22s  n=%4d  nt=%5d  best=%9.1f ms  avg=%9.1f ms%n",
                c.label(), c.n(), c.nt(), best, avg);
            System.out.printf("%s\t%d\t%.1f%n", c.n() + "×" + c.n(), c.nt(), best);
        }
    }
}
